File: android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectDetailsFragment.java
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2012 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc;

import android.app.Activity;
import android.app.Dialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.RemoteException;
import android.text.SpannableString;
import android.text.style.UnderlineSpan;
import android.util.Log;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Button;
import android.widget.HorizontalScrollView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import java.util.ArrayList;
import java.util.List;

import edu.berkeley.boinc.rpc.ImageWrapper;
import edu.berkeley.boinc.rpc.Project;
import edu.berkeley.boinc.rpc.ProjectInfo;
import edu.berkeley.boinc.rpc.RpcClient;
import edu.berkeley.boinc.utils.Logging;

public class ProjectDetailsFragment extends Fragment {
    private String url;
    private ProjectInfo projectInfo; // might be null for projects added via manual URL attach
    private Project project;
    private List<ImageWrapper> slideshowImages = new ArrayList<>();

    private LayoutInflater li;
    private View root;
    private HorizontalScrollView slideshowWrapper;
    private LinearLayout slideshowHook;
    private ProgressBar slideshowLoading;

    // display dimensions
    private int width;
    private int height;

    private boolean retryLayout = true;

    // BroadcastReceiver event is used to update the UI with updated information from
    // the client.  This is generally called once a second.
    //
    private IntentFilter ifcsc = new IntentFilter("edu.berkeley.boinc.clientstatuschange");
    private BroadcastReceiver mClientStatusChangeRec = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            getCurrentProjectData();
            if(retryLayout) {
                populateLayout();
            }
            else {
                updateChangingItems(root);
            }
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        // get data
        url = getArguments().getString("url");
        getCurrentProjectData();

        setHasOptionsMenu(true); // enables fragment specific menu
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        if(Logging.VERBOSE) {
            Log.v(Logging.TAG, "ProjectDetailsFragment onCreateView");
        }
        // Inflate the layout for this fragment
        this.li = inflater;
        View layout = inflater.inflate(R.layout.project_details_layout, container, false);
        root = layout;
        return layout;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        if (context instanceof Activity) {
            final Activity activity = (Activity) context;
            final Display display = activity.getWindowManager().getDefaultDisplay();

            Point size = new Point();
            display.getSize(size);

            width = size.x;
            height = size.y;
        }

        super.onAttach(context);
    }

    @Override
    public void onPause() {
        getActivity().unregisterReceiver(mClientStatusChangeRec);
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        getActivity().registerReceiver(mClientStatusChangeRec, ifcsc);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
        // appends the project specific menu to the main menu.
        inflater.inflate(R.menu.project_details_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {

        super.onPrepareOptionsMenu(menu);
        if(project == null) {
            return;
        }

        // no new tasks, adapt based on status
        MenuItem nnt = menu.findItem(R.id.projects_control_nonewtasks);
        if(project.getDoNotRequestMoreWork()) {
            nnt.setTitle(R.string.projects_control_allownewtasks);
        }
        else {
            nnt.setTitle(R.string.projects_control_nonewtasks);
        }

        // project suspension, adapt based on status
        MenuItem suspend = menu.findItem(R.id.projects_control_suspend);
        if(project.getSuspendedViaGUI()) {
            suspend.setTitle(R.string.projects_control_resume);
        }
        else {
            suspend.setTitle(R.string.projects_control_suspend);
        }

        // detach, only show when project not managed
        MenuItem remove = menu.findItem(R.id.projects_control_remove);
        if(project.getAttachedViaAcctMgr()) {
            remove.setVisible(false);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()) {
            case R.id.projects_control_update:
                new ProjectOperationAsync().execute(RpcClient.PROJECT_UPDATE);
                break;
            case R.id.projects_control_suspend:
                if(project.getSuspendedViaGUI()) {
                    new ProjectOperationAsync().execute(RpcClient.PROJECT_RESUME);
                }
                else {
                    new ProjectOperationAsync().execute(RpcClient.PROJECT_SUSPEND);
                }
                break;
            case R.id.projects_control_nonewtasks:
                if(project.getDoNotRequestMoreWork()) {
                    new ProjectOperationAsync().execute(RpcClient.PROJECT_ANW);
                }
                else {
                    new ProjectOperationAsync().execute(RpcClient.PROJECT_NNW);
                }
                break;
            case R.id.projects_control_reset:
                showConfirmationDialog(RpcClient.PROJECT_RESET);
                break;
            case R.id.projects_control_remove:
                showConfirmationDialog(RpcClient.PROJECT_DETACH);
                break;
            default:
                if(Logging.WARNING) {
                    Log.w(Logging.TAG, "ProjectDetailsFragment onOptionsItemSelected: could not match ID");
                }
        }

        return super.onOptionsItemSelected(item);
    }

    private void showConfirmationDialog(final int operation) {
        final Dialog dialog = new Dialog(getActivity());
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.setContentView(R.layout.dialog_confirm);
        Button confirm = dialog.findViewById(R.id.confirm);
        TextView tvTitle = dialog.findViewById(R.id.title);
        TextView tvMessage = dialog.findViewById(R.id.message);

        // operation-dependent texts
        if(operation == RpcClient.PROJECT_DETACH) {
            final String removeStr = getString(R.string.projects_confirm_detach_confirm);

            tvTitle.setText(getString(R.string.projects_confirm_title, removeStr));
            tvMessage.setText(getString(R.string.projects_confirm_message,
                                        removeStr.toLowerCase(),
                                        project.getProjectName() + " "
                                        + getString(R.string.projects_confirm_detach_message)));
            confirm.setText(removeStr);
        }
        else if(operation == RpcClient.PROJECT_RESET) {
            final String resetStr = getString(R.string.projects_confirm_reset_confirm);

            tvTitle.setText(getString(R.string.projects_confirm_title, resetStr));
            tvMessage.setText(getString(R.string.projects_confirm_message, resetStr.toLowerCase(),
                                        project.getProjectName()));
            confirm.setText(resetStr);
        }

        confirm.setOnClickListener(view -> {
            new ProjectOperationAsync().execute(operation);
            dialog.dismiss();
        });
        Button cancel = dialog.findViewById(R.id.cancel);
        cancel.setOnClickListener(view -> dialog.dismiss());
        dialog.show();
    }

    private void populateLayout() {
        if(project == null) {
            retryLayout = true;
            return; // if data not available yet, return. frequently retrys with onReceive
        }

        retryLayout = false;
        View v = root;

        updateChangingItems(v);

        slideshowWrapper = v.findViewById(R.id.slideshow_wrapper);
        slideshowHook = v.findViewById(R.id.slideshow_hook);
        slideshowLoading = v.findViewById(R.id.slideshow_loading);

        // set website
        TextView website = v.findViewById(R.id.project_url);
        SpannableString content = new SpannableString(project.getMasterURL());
        content.setSpan(new UnderlineSpan(), 0, content.length(), 0);
        website.setText(content);
        website.setOnClickListener(view -> {
            Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(project.getMasterURL()));
            startActivity(i);
        });

        // set general area
        if(projectInfo != null && projectInfo.getGeneralArea() != null) {
            TextView generalArea = v.findViewById(R.id.general_area);
            generalArea.setText(projectInfo.getGeneralArea());
        }
        else {
            LinearLayout wrapper = v.findViewById(R.id.general_area_wrapper);
            wrapper.setVisibility(View.GONE);
        }

        // set specific area
        if(projectInfo != null && projectInfo.getSpecificArea() != null) {
            TextView specificArea = v.findViewById(R.id.specific_area);
            specificArea.setText(projectInfo.getSpecificArea());
        }
        else {
            LinearLayout wrapper = v.findViewById(R.id.specific_area_wrapper);
            wrapper.setVisibility(View.GONE);
        }

        // set description
        if(projectInfo != null && projectInfo.getDescription() != null) {
            TextView description = v.findViewById(R.id.description);
            description.setText(projectInfo.getDescription());
        }
        else {
            LinearLayout wrapper = v.findViewById(R.id.description_wrapper);
            wrapper.setVisibility(View.GONE);
        }

        // set home
        if(projectInfo != null && projectInfo.getHome() != null) {
            TextView home = v.findViewById(R.id.based_at);
            home.setText(projectInfo.getHome());
        }
        else {
            LinearLayout wrapper = v.findViewById(R.id.based_at_wrapper);
            wrapper.setVisibility(View.GONE);
        }

        // load slideshow
        new UpdateSlideshowImagesAsync().execute();
    }

    private void getCurrentProjectData() {
        try {
            List<Project> allProjects = (ArrayList<Project>) BOINCActivity.monitor.getProjects();
            for(Project tmpP : allProjects) {
                if(tmpP.getMasterURL().equals(url)) {
                    this.project = tmpP;
                }
            }
            this.projectInfo = BOINCActivity.monitor.getProjectInfo(url);
        }
        catch(Exception e) {
            if(Logging.ERROR) {
                Log.e(Logging.TAG, "ProjectDetailsFragment getCurrentProjectData could not retrieve project list");
            }
        }
        if(this.project == null) {
            if(Logging.WARNING) {
                Log.w(Logging.TAG,
                      "ProjectDetailsFragment getCurrentProjectData could not find project for URL: " + url);
            }
        }
        if(this.projectInfo == null) {
            if(Logging.DEBUG) {
                Log.d(Logging.TAG,
                      "ProjectDetailsFragment getCurrentProjectData could not find project attach list for URL: " +
                      url);
            }
        }
    }

    private void updateChangingItems(View v) {
        try {
            // status
            String newStatus = BOINCActivity.monitor.getProjectStatus(project.getMasterURL());
            LinearLayout wrapper = v.findViewById(R.id.status_wrapper);
            if(!newStatus.isEmpty()) {
                wrapper.setVisibility(View.VISIBLE);
                TextView statusT = v.findViewById(R.id.status_text);
                statusT.setText(newStatus);
            }
            else {
                wrapper.setVisibility(View.GONE);
            }

        }
        catch(Exception e) {
            if(Logging.ERROR) {
                Log.e(Logging.TAG, "ProjectDetailsFragment.updateChangingItems error: ", e);
            }
        }
    }

    // executes project operations in new thread
    private final class ProjectOperationAsync extends AsyncTask<Object, Void, Boolean> {

        @Override
        protected Boolean doInBackground(Object... params) {
            if(Logging.DEBUG) {
                Log.d(Logging.TAG, "ProjectOperationAsync doInBackground");
            }
            try {
                Integer operation = (Integer) params[0];
                return BOINCActivity.monitor.projectOp(operation, project.getMasterURL());
            }
            catch(Exception e) {
                if(Logging.WARNING) {
                    Log.w(Logging.TAG, "ProjectOperationAsync error in do in background", e);
                }
            }
            return false;
        }

        @Override
        protected void onPostExecute(Boolean success) {
            if(success) {
                try {
                    BOINCActivity.monitor.forceRefresh();
                }
                catch(RemoteException e) {
                    if(Logging.ERROR) {
                        Log.e(Logging.TAG, "ProjectDetailsFragment.ProjectOperationAsync.onPostExecute() error: ", e);
                    }
                }
            }
            else if(Logging.WARNING) {
                Log.w(Logging.TAG, "ProjectOperationAsync failed.");
            }
        }
    }

    private final class UpdateSlideshowImagesAsync extends AsyncTask<Void, Void, Boolean> {

        @Override
        protected Boolean doInBackground(Void... params) {
            if(Logging.DEBUG) {
                Log.d(Logging.TAG,
                      "UpdateSlideshowImagesAsync updating images in new thread. project: " +
                      project.getMasterURL());
            }
            try {
                slideshowImages =
                        BOINCActivity.monitor.getSlideshowForProject(project.getMasterURL());
            }
            catch(Exception e) {
                if(Logging.WARNING) {
                    Log.w(Logging.TAG, "UpdateSlideshowImagesAsync: Could not load data, clientStatus not initialized.");
                }
                return false;
            }
            return !slideshowImages.isEmpty();
        }

        @Override
        protected void onPostExecute(Boolean success) {
            if(Logging.DEBUG) {
                Log.d(Logging.TAG,
                      "UpdateSlideshowImagesAsync success: " + success + " images: " + slideshowImages.size());
            }

            if(success && !slideshowImages.isEmpty()) {
                slideshowLoading.setVisibility(View.GONE);
                for(ImageWrapper image : slideshowImages) {
                    ImageView iv = (ImageView) li.inflate(R.layout.project_details_slideshow_image_layout, null);
                    Bitmap bitmap = image.getImage();
                    assert bitmap != null;
                    if(scaleImages(bitmap.getHeight(), bitmap.getWidth())) {
                        bitmap = Bitmap.createScaledBitmap(image.getImage(),
                                                           image.getImage().getWidth() * 2,
                                                           image.getImage().getHeight() * 2, false);
                    }
                    iv.setImageBitmap(bitmap);
                    slideshowHook.addView(iv);
                }
            }
            else {
                slideshowWrapper.setVisibility(View.GONE);
            }
        }

        private boolean scaleImages(int imageHeight, int imageWidth) {
            return (height >= imageHeight * 2 && width >= imageWidth * 2);
        }
    }
}

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectDetailsFragment.kt
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2021 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc

import android.app.Activity
import android.app.Dialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Point
import android.os.Build
import android.os.Bundle
import android.os.RemoteException
import android.text.SpannableString
import android.text.style.UnderlineSpan
import android.view.*
import android.widget.Button
import android.widget.TextView
import androidx.core.graphics.scale
import androidx.core.net.toUri
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import edu.berkeley.boinc.databinding.ProjectDetailsLayoutBinding
import edu.berkeley.boinc.databinding.ProjectDetailsSlideshowImageLayoutBinding
import edu.berkeley.boinc.rpc.ImageWrapper
import edu.berkeley.boinc.rpc.Project
import edu.berkeley.boinc.rpc.ProjectInfo
import edu.berkeley.boinc.rpc.RpcClient
import edu.berkeley.boinc.utils.Logging
import java.util.*
import kotlinx.coroutines.*

class ProjectDetailsFragment : Fragment() {
    private var url: String = ""

    // might be null for projects added via manual URL attach
    private var projectInfo: ProjectInfo? = null

    private var project: Project? = null
    private var slideshowImages: List<ImageWrapper> = ArrayList()

    private var _binding: ProjectDetailsLayoutBinding? = null
    private val binding get() = _binding!!

    // display dimensions
    private var width = 0
    private var height = 0
    private var retryLayout = true

    private val currentProjectData: Unit
        get() {
            try {
                project = BOINCActivity.monitor!!.projects.firstOrNull { it.masterURL == url }
                projectInfo = BOINCActivity.monitor!!.getProjectInfoAsync(url).await()
            } catch (e: Exception) {
                Logging.logError(Logging.Category.GUI_VIEW, "ProjectDetailsFragment getCurrentProjectData could not" +
                        " retrieve project list")
            }
            if (project == null) {
                Logging.logWarning(Logging.Category.GUI_VIEW,
                        "ProjectDetailsFragment getCurrentProjectData could not find project for URL: $url")
            }
            if (projectInfo == null) {
                Logging.logDebug(Logging.Category.GUI_VIEW,
                        "ProjectDetailsFragment getCurrentProjectData could not find project" +
                        " attach list for URL: $url")
            }
        }

    // BroadcastReceiver event is used to update the UI with updated information from
    // the client.  This is generally called once a second.
    //
    private val ifcsc = IntentFilter("edu.berkeley.boinc.clientstatuschange")
    private val mClientStatusChangeRec: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            currentProjectData
            if (retryLayout) {
                populateLayout()
            } else {
                updateChangingItems()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        // get data
        url = requireArguments().getString("url") ?: ""
        currentProjectData

        super.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val menuHost: MenuHost = requireActivity() // enables fragment specific menu

        // add the project menu to the fragment
        menuHost.addMenuProvider(object: MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.project_details_menu, menu)
            }

            override fun onPrepareMenu(menu: Menu) {
                super.onPrepareMenu(menu)

                if (project == null) {
                    return
                }

                // no new tasks, adapt based on status
                val nnt = menu.findItem(R.id.projects_control_nonewtasks)
                if (project!!.doNotRequestMoreWork) {
                    nnt.setTitle(R.string.projects_control_allownewtasks)
                } else {
                    nnt.setTitle(R.string.projects_control_nonewtasks)
                }

                // project suspension, adapt based on status
                val suspend = menu.findItem(R.id.projects_control_suspend)
                if (project!!.suspendedViaGUI) {
                    suspend.setTitle(R.string.projects_control_resume)
                } else {
                    suspend.setTitle(R.string.projects_control_suspend)
                }

                // detach, only show when project not managed
                val remove = menu.findItem(R.id.projects_control_remove)
                if (project!!.attachedViaAcctMgr) {
                    remove.isVisible = false
                }
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                lifecycleScope.launch {
                    when (menuItem.itemId) {
                        R.id.projects_control_update -> performProjectOperation(RpcClient.PROJECT_UPDATE)
                        R.id.projects_control_suspend -> if (project!!.suspendedViaGUI) {
                            performProjectOperation(RpcClient.PROJECT_RESUME)
                        } else {
                            performProjectOperation(RpcClient.PROJECT_SUSPEND)
                        }
                        R.id.projects_control_nonewtasks -> if (project!!.doNotRequestMoreWork) {
                            performProjectOperation(RpcClient.PROJECT_ANW)
                        } else {
                            performProjectOperation(RpcClient.PROJECT_NNW)
                        }
                        R.id.projects_control_reset -> showConfirmationDialog(RpcClient.PROJECT_RESET)
                        R.id.projects_control_remove -> showConfirmationDialog(RpcClient.PROJECT_DETACH)
                        else -> {
                            Logging.logError(Logging.Category.USER_ACTION, "ProjectDetailsFragment onOptionsItemSelected: could not match ID")
                        }
                    }
                }

                return true
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        Logging.logVerbose(Logging.Category.GUI_VIEW, "ProjectDetailsFragment onCreateView")

        // Inflate the layout for this fragment
        _binding = ProjectDetailsLayoutBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onAttach(context: Context) {
        if (context is Activity) {
            val size = Point()
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
                @Suppress("DEPRECATION")
                context.windowManager.defaultDisplay.getSize(size)
            } else {
                val r = context.windowManager.currentWindowMetrics.bounds
                size.x = r.width()
                size.y = r.height()
            }
            width = size.x
            height = size.y
        }
        super.onAttach(context)
    }

    override fun onPause() {
        activity?.unregisterReceiver(mClientStatusChangeRec)
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        activity?.registerReceiver(mClientStatusChangeRec, ifcsc)
    }

    private fun showConfirmationDialog(operation: Int) {
        val dialog = Dialog(requireActivity()).apply {
            requestWindowFeature(Window.FEATURE_NO_TITLE)
            setContentView(R.layout.dialog_confirm)
        }
        val confirm = dialog.findViewById<Button>(R.id.confirm)
        val tvTitle = dialog.findViewById<TextView>(R.id.title)
        val tvMessage = dialog.findViewById<TextView>(R.id.message)

        // operation-dependent texts
        if (operation == RpcClient.PROJECT_DETACH) {
            val removeStr = getString(R.string.projects_confirm_detach_confirm)
            tvTitle.text = getString(R.string.projects_confirm_title, removeStr)
            tvMessage.text = getString(R.string.projects_confirm_message,
                removeStr.lowercase(Locale.ROOT), project!!.projectName + " "
                            + getString(R.string.projects_confirm_detach_message))
            confirm.text = removeStr
        } else if (operation == RpcClient.PROJECT_RESET) {
            val resetStr = getString(R.string.projects_confirm_reset_confirm)
            tvTitle.text = getString(R.string.projects_confirm_title, resetStr)
            tvMessage.text = getString(R.string.projects_confirm_message,
                resetStr.lowercase(Locale.ROOT),
                    project!!.projectName)
            confirm.text = resetStr
        }
        confirm.setOnClickListener {
            lifecycleScope.launch {
                performProjectOperation(operation)
            }
            dialog.dismiss()
        }
        dialog.findViewById<Button>(R.id.cancel).apply {
            setOnClickListener { dialog.dismiss() }
        }
        dialog.show()
    }

    private fun populateLayout() {
        if (project == null) {
            retryLayout = true
            return  // if data not available yet, return. Frequently retry with onReceive
        }
        retryLayout = false
        updateChangingItems()

        // set website
        val content = SpannableString(project!!.masterURL)
        content.setSpan(UnderlineSpan(), 0, content.length, 0)
        binding.projectUrl.text = content
        binding.projectUrl.setOnClickListener {
            startActivity(Intent(Intent.ACTION_VIEW, project!!.masterURL.toUri()))
        }

        // set general area
        if (projectInfo?.generalArea != null) {
            binding.generalArea.text = projectInfo!!.generalArea
        } else {
            binding.generalAreaWrapper.visibility = View.GONE
        }

        // set specific area
        if (projectInfo?.specificArea != null) {
            binding.specificArea.text = projectInfo!!.specificArea
        } else {
            binding.specificAreaWrapper.visibility = View.GONE
        }

        // set description
        if (projectInfo?.description != null) {
            binding.description.text = projectInfo!!.description
        } else {
            binding.descriptionWrapper.visibility = View.GONE
        }

        // set home
        if (projectInfo?.home != null) {
            binding.basedAt.text = projectInfo!!.home
        } else {
            binding.basedAtWrapper.visibility = View.GONE
        }

        // load slideshow
        lifecycleScope.launch {
            updateSlideshowImages()
        }
    }

    private fun updateChangingItems() {
        try {
            // status
            val newStatus = BOINCActivity.monitor!!.getProjectStatus(project!!.masterURL)
            if (newStatus.isNotEmpty()) {
                binding.statusWrapper.visibility = View.VISIBLE
                binding.statusText.text = newStatus
            } else {
                binding.statusWrapper.visibility = View.GONE
            }
        } catch (e: Exception) {
            Logging.logException(Logging.Category.GUI_VIEW, "ProjectDetailsFragment.updateChangingItems error: ", e)
        }
    }

    // executes project operations in new thread
    private suspend fun performProjectOperation(operation: Int) = coroutineScope {
        Logging.logVerbose(Logging.Category.USER_ACTION, "performProjectOperation()")

        val success = async {
            return@async try {
                BOINCActivity.monitor!!.projectOp(operation, project!!.masterURL)
            } catch (e: Exception) {
                Logging.logException(Logging.Category.USER_ACTION, "performProjectOperation() error: ", e)

                false
            }
        }.await()

        if (success) {
            try {
                BOINCActivity.monitor!!.forceRefresh()
            } catch (e: RemoteException) {
                Logging.logException(Logging.Category.USER_ACTION, "performProjectOperation() error: ", e)
            }
        } else {
            Logging.logError(Logging.Category.USER_ACTION, "performProjectOperation() failed.")
        }
    }

    private suspend fun updateSlideshowImages() = coroutineScope {
        Logging.logDebug(Logging.Category.GUI_VIEW,
                "UpdateSlideshowImagesAsync updating images in new thread. project:" +
                " $project!!.masterURL")

        val success = withContext(Dispatchers.Default) {
            slideshowImages = try {
                BOINCActivity.monitor!!.getSlideshowForProject(project!!.masterURL)
            } catch (e: Exception) {
                Logging.logError(Logging.Category.GUI_VIEW, "updateSlideshowImages: Could not load data, " +
                        "clientStatus not initialized.")

                return@withContext false
            }
            return@withContext slideshowImages.isNotEmpty()
        }

        Logging.logDebug(Logging.Category.GUI_VIEW,
                "UpdateSlideshowImagesAsync success: $success, images: ${slideshowImages.size}")
        if (success && slideshowImages.isNotEmpty()) {
            binding.slideshowLoading.visibility = View.GONE
            for (image in slideshowImages) {
                val slideshowBinding = ProjectDetailsSlideshowImageLayoutBinding.inflate(layoutInflater)
                var bitmap = image.image!!
                if (scaleImages(bitmap.height, bitmap.width)) {
                    bitmap = bitmap.scale(bitmap.width * 2, bitmap.height * 2, filter = false)
                }
                slideshowBinding.slideshowImage.setImageBitmap(bitmap)
                binding.slideshowHook.addView(slideshowBinding.slideshowImage)
            }
        } else {
            binding.slideshowWrapper.visibility = View.GONE
        }
    }

    private fun scaleImages(imageHeight: Int, imageWidth: Int) = height >= imageHeight * 2 && width >= imageWidth * 2
}
